restli-server/src/main/java/com/linkedin/restli/server/annotations/Action.java

```diff
@@ -47,6 +47,15 @@
    */
   ResourceLevel resourceLevel() default ResourceLevel.ANY;
 
+  /**
+   * Optional placeholder attribute used to indicate whether the action is read-only
+   * or not. Default is false which indicates that this action can result in writes. This
+   * parameter is a placeholder to provide a hint to metadata introspection systems. The
+   * framework itself does not and cannot actually enforce read-only behavior in actions
+   * annotated as such.
+   */
+  boolean readOnly() default false;
```

Comment:
Have you verified that it will appear in IDL? I suspect that this is not enough, and we may need to make change in ActionSchema used in ResourceModel to export to IDL so that `graphql-infra` can read from to get `ResourceSchema`.

Reply:
Adding it here only will not exposed it in the IDL. That's the reason why I asked what's the way to consume this hint? If it will consumed from IDL, then ActionSchema, RestliAnnotationReader etc need to be updated to make it is exposed in the IDL.

Reply:
yeah, GraphQL infra needs to consume it from ResourceSchema to decide whether to generate this under Mutation or under Query. Of course, it can also be done through reading custom annotation (works but not very friendly). We can add it into ActionSchema to make it first-class citizen property.

Reply:
> I am skeptical about whether this needs to be defined as a first-class property on the `Action` annotation. In fact, I don't think it's appropriate because the native Rest.li framework doesn't act upon this property. Why are custom Rest.li annotations not sufficient in this case? That would be much less intrusive.

rest.li compatibility checker needs to report warning if the value is changed, does this count as "rest.li framework acts upon this property"?

Reply:
@mchen07 Correct. If there will be behavior natively supported by the framework corresponding to this attribute (e.g. compatibility checker logic), then supporting the attribute as a first-class feature is warranted. Otherwise, implementing this as a decoupled extension (e.g. custom rest annotation) is more reasonable.

I think if we implement corresponding compatibility checker logic, then including `readOnly` as a first-class feature is okay.

Reply:
Ok added it to IDL, and added backward compat check for it. Making something readwrite as readonly, or vice-versa is a backward incompatible change.

The reason is that if something is read-write, it will appear as a GQL mutation, and if read-only as a query.